klippy/extras/pwm_tool.py

```diff
@@ -74,39 +86,50 @@ def _build_config(self):
         # Software PWM
         if self._shutdown_value not in [0., 1.]:
             raise pins.error("shutdown value must be 0.0 or 1.0 on soft pwm")
-        if cycle_ticks >= 1<<31:
-            raise pins.error("PWM pin cycle time too large")
         self._mcu.add_config_cmd(
             "config_digital_out oid=%d pin=%s value=%d"
             " default_value=%d max_duration=%d"
             % (self._oid, self._pin, self._start_value >= 1.0,
-               self._shutdown_value >= 0.5, mdur_ticks))
+               self._shutdown_value >= 0.5, self._duration_ticks))
+        self._default_value = int(self._shutdown_value >= 0.5) * cycle_ticks
         self._mcu.add_config_cmd(
             "set_digital_out_pwm_cycle oid=%d cycle_ticks=%d"
             % (self._oid, cycle_ticks))
-        self._cycle_ticks = cycle_ticks
-        svalue = int(self._start_value * cycle_ticks + 0.5)
+        self._pwm_max = float(cycle_ticks)
+        self._last_value = int(self._start_value * self._pwm_max + 0.5)
         self._mcu.add_config_cmd(
             "queue_digital_out oid=%d clock=%d on_ticks=%d"
-            % (self._oid, self._last_clock, svalue), is_init=True)
+            % (self._oid, self._last_clock, self._last_value), is_init=True)
         self._set_cmd_tag = self._mcu.lookup_command(
             "queue_digital_out oid=%c clock=%u on_ticks=%u",
             cq=cmd_queue).get_command_tag()
-    def set_pwm(self, print_time, value):
-        clock = self._mcu.print_time_to_clock(print_time)
-        minclock = self._last_clock
-        self._last_clock = clock
-        if self._invert:
-            value = 1. - value
-        max_count = self._cycle_ticks
-        if self._hardware_pwm:
-            max_count = self._pwm_max
-        v = int(max(0., min(1., value)) * max_count + 0.5)
-        data = (self._set_cmd_tag, self._oid, clock & 0xffffffff, v)
+    def _send_update(self, clock, val):
+        self._last_clock = clock = max(self._last_clock, clock)
+        self._last_value = val
+        data = (self._set_cmd_tag, self._oid, clock & 0xffffffff, val)
         ret = self._stepcompress_queue_mq_msg(self._stepqueue, clock,
                                               data, len(data))
         if ret:
             raise error("Internal error in stepcompress")
+        # Notify toolhead so that it will flush this update
+        wakeclock = clock
+        if self._last_value != self._default_value:
+            # Continue flushing to resend time
+            wakeclock += self._duration_ticks
+        wake_print_time = self._mcu.clock_to_print_time(wakeclock)
+        self._toolhead.note_kinematic_activity(wake_print_time
+                                               + CLOCK_SKEW_EXTRA_TIME)
```

Comment:
`SKEW_EXTRA_TIME` is the buffer time to allow latency between MCU and host?

Reply:
It's to account for minor changes in clock synchronization.  If we need to flush to clock 800000, which translates to time 100.0, but when time 100.0 actually comes around if an updated clock synchronization translates that to 799998, then an update at 800000 wont get flushed.

Basically, there is no harm in leaving flushing enabled for longer than necessary, so it's easier to add a little bit of extra time than to risk going under.

-Kevin

Reply:
Ah, so to flush further into "the future", I understand now.